File: java/Scrape.java
// Fetch a page through the ScrapingBee web scraping API.
// Auth is the Authorization: Bearer header. The api_key query parameter is deprecated.
//
// Java 11+ runs this file directly, no build tool and no dependencies:
//   java Scrape.java

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

public class Scrape {

    private static final String API_URL = "https://app.scrapingbee.com/api/v1";

    private static String apiKey() {
        String key = System.getenv("SCRAPINGBEE_API_KEY");
        return (key == null || key.isEmpty()) ? "YOUR-API-KEY" : key;
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    static HttpResponse<String> scrape(String target, String extraParams) throws Exception {
        String query = "?url=" + encode(target) + extraParams;

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(API_URL + query))
                .header("Authorization", "Bearer " + apiKey())
                .GET()
                .build();

        return HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString());
    }

    public static void main(String[] args) throws Exception {
        // render_js=false keeps this call at 1 credit instead of the 5-credit default.
        HttpResponse<String> response = scrape("https://news.ycombinator.com", "&render_js=false");

        System.out.println("status: " + response.statusCode());
        System.out.println("credits: " + response.headers().firstValue("Spb-cost").orElse("n/a"));
        System.out.println("request id: " + response.headers().firstValue("Spb-request-id").orElse("n/a"));

        String body = response.body();
        System.out.println(body.substring(0, Math.min(500, body.length())));
    }
}

File: php/scrape.php
<?php
// Fetch a page through the ScrapingBee web scraping API.
// Auth is the Authorization: Bearer header. The api_key query parameter is deprecated.

const API_URL = 'https://app.scrapingbee.com/api/v1';

function scrape(string $target, array $params = []): array
{
    $apiKey = getenv('SCRAPINGBEE_API_KEY') ?: 'YOUR-API-KEY';
    $query = http_build_query(array_merge(['url' => $target], $params));

    $ch = curl_init(API_URL . '?' . $query);
    curl_setopt($ch, CURLOPT_HTTPHEADER, ['Authorization: Bearer ' . $apiKey]);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HEADER, true);

    $raw = curl_exec($ch);
    if ($raw === false) {
        throw new RuntimeException('Request failed: ' . curl_error($ch));
    }

    $headerSize = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
    $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    return [
        'status' => $status,
        'headers' => substr($raw, 0, $headerSize),
        'body' => substr($raw, $headerSize),
    ];
}

// render_js=false keeps this call at 1 credit instead of the 5-credit default.
$response = scrape('https://news.ycombinator.com', ['render_js' => 'false']);

echo 'status: ' . $response['status'] . PHP_EOL;

// HTTP/2 lowercases header names, so match case-insensitively.
preg_match_all('/^spb-[^:]+:.*$/im', $response['headers'], $matches);
echo implode(PHP_EOL, $matches[0]) . PHP_EOL;

echo substr($response['body'], 0, 500) . PHP_EOL;
